docs/madr/decisions/040-meshservice-hostname-vips.md

```diff
@@ -0,0 +1,140 @@
+  The current value of 30s is too long, especially that the DNS server is run locally in kuma-dp. Leader election should take more than X seconds, so if we were to lose this list, it’s not a problem.
+
+Following this process, we don’t have to store VIPs in any central place like we do currently (VIP ConfigMap for each Mesh).
+We also assign VIPs for ExternalServices, but those can take a separate CIDR to avoid collisions.
+
+To avoid collision with an existing VIPs mechanism, we can take a completely new CIDR by default (241.0.0.0/4).
+This CIDR can be configured by the user.
+
+### Hostname Generator
+
+MeshService has to be addressable with a hostname.
+In case of services in a Kubernetes cluster, we don't have to do anything for local services because we can just rely on Kube DNS that resolves to a Kube VIP we placed for MeshService.
+
+In case of services in a Universal cluster, currently we provide two ways:
+* `<kuma.io/service>.mesh` address that resolves to Kuma VIP
+* VirtualOutbound resource to provide custom addresses
+
+Hostnames should be managed only by MeshOperator. Otherwise, we get into security issues such as one service owner can override a hostname of other service.
+Hostnames are created by convention in the majority of cases - Kubernetes hostnames are also such example.
+It is still possible for service owner to create unique hostname for their app by reaching out to mesh operator.
+
+To manage hostname, we introduce new object called `HostnameGenerator`
+
+```yaml
+type: HostnameGenerator
+name: k8s-zone-hostnames
+spec:
+  selector:
+    meshService:
+      Labels:
+        kuma.io/zone: east 
+  template: {{ label "k8s.kuma.io/service-name" }}.{{ label "k8s.kuma.io/namespace" }}.svc.mesh.{{ label "kuma.io/zone" }}
+```
+
+`HostnameGenerator` is a namespaced-scoped object on Kubernetes, but can only be applied in system namespace.
+It is namespaced-scoped object, because we already learned that it's better to start with restricted namespace object than to start with cluster-scope and then make it namespaced-scope.
+We might need this for deploying Global and Zone CP in one Kube cluster and maybe potentially enable applying generator on namespaces to affect only data plane proxes in specific namespace.
+It can be applied on global CP, so it's synced to all zones, or it can be applied on specific zones.
+
+For example, given this mesh service
+
+```yaml
+type: MeshService
+name: redis.demo-app
+labels:
+  k8s.kuma.io/service-name: redis
+  k8s.kuma.io/namespace: demo-app
+  kuma.io/zone: east
+spec: ...
+```
+We would generate such hostname `redis.demo-app.svc.mesh.east`
+
+Possible template functions/keys are:
+* `{{ name }}` - name of the MeshService
+* `{{ label "x" }}` - value of label `x`.
+
+If the template cannot be resolved (label is missing), the hostname won't be generated with a `NotAvailable` status and an appropriate `reason`.
```

Comment:
hostname in addresses would be empty then?

The status would look like:

```yaml
status:
  addresses:
    - hostname: ""
      status: NotAvailable # if there was a collision
      origin:
        kind: HostnameGenerator
        name: "k8s-zone-hostnames" # name of the HostnameGenerator
      reason: "not available because of the clash with ..." # reason if there was a problem
```

Reply:
I think we would just skip the template. Imagine you have this template
```
{{ label "kuma.io/display-name" }}.{{ label "k8s.kuma.io/namespace" }}.svc.mesh.{{ label "k8s.kuma.io/zone" }}
```
You have a universal zone that syncs the service to you that does not have such labels.
You can of course select all kubernetes zones one by one for this generator, but that's annoying.